crates/stages/src/stage.rs

```diff
@@ -65,6 +87,29 @@ pub trait Stage<DB: Database>: Send + Sync {
     /// Stage IDs must be unique.
     fn id(&self) -> StageId;
 
+    /// Return next execution action.
+    ///
+    /// [ExecAction::Done] is returned if there are no blocks to execute in this stage.
+    /// [ExecAction::Run] is returned if the stage should proceed with execution.
+    fn next_exec_action(&self, input: &ExecInput, max_threshold: Option<u64>) -> ExecAction {
+        // Extract information about the stage progress
+        let stage_progress = input.stage_progress.unwrap_or_default();
+        let previous_stage_progress = input.previous_stage_progress();
+
+        let start_block = stage_progress + 1;
+        let end_block = match max_threshold {
+            Some(threshold) => previous_stage_progress.min(stage_progress + threshold),
+            None => previous_stage_progress,
+        };
+        let capped = end_block < previous_stage_progress;
```

Comment:
I find `capped` a bit confusing, maybe there's a better name for it?

Should we package `commit_threshold` as part of the `ExecInput` and add a `ExecInput::new(...)` function which does any sanity checks?

And then add a `ExecInput::next(mut self)` method which does all the logic in `next_exec_action`?

That would let us call `input.next()` inside of `execute`.

We could make it even simpler via a `let done = run_or_continue_stage()!` macro which does `input.next()` and the matching, to avoid the matching downstream.

Reply:
@gakonst what's the better name for `capped`?

Reply:
the problem with moving the threshold to the input is that it's specific to each stage. imo it would be a bit dirty to query the stage for the threshold only to provide it as input later